---
title: "MB194 C.tr infection primary cervical cells - MicroArray QC"
author: "Hilmar Berger"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---

```{r, message=FALSE}
rm(list=ls())
library(limma)
suppressMessages(library(xlsx))
library(pheatmap)
library(xtable)
library(knitr)
setwd("/data_genome2/projects/MB194_Cervical_organoids/Microarray/Analysis_Infection/")
```

# Data QC and preprocessing for microarray data

```{r, message=FALSE}
ed = read.xlsx("ExpDesign.xls", 1, stringsAsFactors=F)

ed = subset(ed, Included_in_Ctr_publication==1)

ed$sample_ID=as.character(ed$Number)
ed$HPV = gsub("no","noE6E7", ed$HPV)
ed$ShortID = paste(ed$Patient, ed$cells, ed$Infection,ed$Time.p.i., ed$Replicate, ed$HPV,ed$Site, sep="_")
ed$Condition = apply(ed[, c('cells','Infection','Time.p.i.','HPV','Site')], 1, paste, collapse="-")
rownames(ed) = gsub(".txt","",ed$filename)

setwd("../Raw/")
agilent.datacolumns=list(E='gMedianSignal',Eb = 'gBGMedianSignal',isNonUniform='gIsFeatNonUnifOL', isPopOutlier='gIsFeatPopnOL', multDetrendSignal='gMultDetrendSignal');
intensities =read.maimages(paste(ed$filename,sep=""), source="agilent.median", green.only=TRUE, columns=agilent.datacolumns)
setwd("../Analysis_Infection/")

# fix outdated chip annotations
new_anno_file = "/data_genome1/References/MicroArray/Agilent/8x60k_048908_human/Custom_60k_annotations_2016-11-08.Rdata"
load(new_anno_file)
old_anno = intensities$genes
take_over_cols = colnames(old_anno)[!colnames(old_anno) %in% c("GeneName","Description","SystematicName")]
tmp = old_anno[,take_over_cols]
tmp$index=1:nrow(tmp)
tmp = merge(tmp, anno_tab_final, by="ProbeName", all.x=T, sort=F)
new_col_order = c(take_over_cols, colnames(tmp)[!colnames(tmp) %in% take_over_cols])
new_col_order = new_col_order[!new_col_order %in% c("GO_BP","GO_CC","GO_MF")]
new_anno = tmp[order(tmp$index),new_col_order]

intensities$genes = new_anno
```


```{r, message=FALSE, results='hide'}
intensities$probe_exclude=(intensities$isNonUniform>0)|(intensities$isPopOutlier>0)
intensities$E[intensities$probe_exclude] <- NA
```

## Data overview
This document describes the preprocessing and QC of microarray data from MB194 Human Cervix samples with/without infection with C.trachomatis (Rajendra Kumar, Cindrilla Chumduri, Stefanie Koster). 
Micro arrays used had design 048908_D (custom 60K based on Sure G3 v2 + lincRNA from v1 and v2, defined by H.Mollenkopf, MPIIB). 

### Samples 
```{r, results='asis', warning=FALSE}
cols_to_exclude = c('SampleName','Medium','FSK','PassageNo','Included.in.EctoEndoPaper','Harvested','sample_ID')
ed_tab = ed[,!colnames(ed) %in% cols_to_exclude]
rownames(ed_tab) <-NULL
#print.xtable(xtable(ed_tab,display=rep("s",ncol(ed)-length(cols_to_exclude)+1), align=paste(c(rep("|l",ncol(ed)-length(cols_to_exclude)+1),"|"), collapse="")), type="html", file="" , include.rownames=F)
kable(ed_tab)
```

### Conditions

Experiments were done in two batches. The first batch was generated in 2015, the second batch was generated by in 2017/2018 and hybridized in 2018-07. 

```{r}
cfreq = as.data.frame(table(ed$Condition))
colnames(cfreq) = c("Condition", "Count")
cfreq
```


## Raw intensity data

### Excluded probes (non-uniform feature or population outlier)

```{r, fig.width=10}
par(mar=c(11,4,4,2))
ex_cnt = apply(intensities$probe_exclude,2,sum)
barplot(ex_cnt, las=2, names.arg = ed[names(ex_cnt), "ShortID"], main = "number of excluded probes/sample")
```

### Intensity distribution across samples
```{r, boxplot_raw, fig.width=10, fig.height=10}
par(mfrow=c(2,1), mar=c(14,4,4,1))
ii = log2(intensities$E)
colnames(ii) = ed[colnames(ii),]$ShortID
boxplot(ii, las=2, main = "Raw FG intensities", ylim=c(0,20))
ii = log2(intensities$Eb)
colnames(ii) = ed[colnames(ii),]$ShortID
boxplot(ii, las=2, main = "Raw BG intensities", ylim=c(0,20))
par(mfrow=c(1,1))
plotDensities(intensities, group = ed[colnames(intensities$E),"ShortID"], legend="topright", main="Intensity distribution by experiment run ID")

```

### Correlation of raw intensities across samples

```{r, heatmap_cor_raw, fig.width=12, fig.height=10}
raw_cor = cor(log2(intensities$E),method="spearman", use="pairwise")
colnames(raw_cor) = ed[colnames(raw_cor),]$ShortID
#rownames(raw_cor) = ed[rownames(raw_cor),]$sample_ID
pheatmap(raw_cor, cluster_rows = F, cluster_cols=F)
```

## Preprocessing

Data will be preprocessed as follows:

* Background for each array will be corrected using the "normexp" method (Ritchie et al 2007, Bioinformatics, p. 2700-07) and an offset of 15. 
* Between-array normalization will be perfomed using the quantile method (Bolstad et al 2003, Bioinformatics, p. 185).

## Background correction
```{r, bg_correction, results='hide'}
bg_corrected = backgroundCorrect(intensities,method="normexp",offset=15)
```

### Intensity distribution across samples after background correction
```{r, boxplot_fg_bg_corrected, fig.width=10}
par(mar=c(14,4,4,1))
boxplot(log2(bg_corrected$E), las=2,  names = ed[colnames(bg_corrected$E), "ShortID"], main = "BG-corrected FG intensities", ylim=c(-2,20))
```


### Normalization

```{r, normalization, echo=TRUE, results='hide'}
normalized = normalizeBetweenArrays(bg_corrected,method="quantile")
```

#### Intensity distribution across samples after BG correction and normalization
```{r, boxplot_fg_normalized, fig.width=10}
par(mar=c(11,4,4,1))
boxplot(log2(normalized$E), las=2, names = ed[colnames(normalized$E), "ShortID"],main = "Normalized FG intensities")
```

```{r, fig.width=10, fig.height=8}
plotDensities(normalized, group = ed[colnames(normalized$E),"ShortID"], legend="topright", main="Intensity distribution by experiment run ID")
```


#### Correlation of normalized intensities across samples
```{r, heatmap_cor_normalized, fig.width=12, fig.height=8}
norm_cor = cor(log2(normalized$E),method="spearman", use="pairwise")
colnames(norm_cor) = ed[colnames(norm_cor),]$sample_ID
rownames(norm_cor) = ed[rownames(norm_cor),]$ShortID
pheatmap(norm_cor)
```

## Primary Component Analysis
```{r, PCA, fig.width=8, fig.height=8}
norm_exp = normalized$E
NA_rows = apply(norm_exp,1,function(x) sum(is.na(x)))
pca = prcomp(t(norm_exp[NA_rows==0,]))
imp = summary(pca)$importance

gg = as.factor(apply(ed[colnames(normalized$E), c('cells','Infection','Time.p.i.','HPV','Site')], 1, paste, collapse="-"))
cp = palette(rainbow(length(levels(gg))))
plot(pca$x[,1], pca$x[,2], type="p", xlab=paste("1st PC (",round(imp[2,1],2)*100 ,"% var explained)",sep="" ), ylab=paste("2nd PC (",round(imp[2,2],2)*100 ,"% var explained)",sep="" ), main="PCA on normalized expression data", xlim=c(min(pca$x[,1])*1.5,max(pca$x[,1])*4))
text(pca$x[,1],pca$x[,2],labels=ed[colnames(normalized$E),]$ShortID, col=cp[as.numeric(gg)], adj=-0.05)
abline(h=0)
abline(v=0)
legend("topright", legend=levels(gg), fill=as.numeric(factor(levels(gg), levels=levels(gg))))
```

## Control probes

The following control probes exist on the arrays used in this experiment:

* Corner associated (used for orientation purposes during scanning)
  * Bright corner 
  * Dark corner 
* Negative controls
  * 3xSLv1 (hairpin probe that does not hybridize well with any possible RNA)
* Positive controls
  * Human GAPDH and PGK1 probes
  * Deletion stringency probes (DCP, probe with varying number of insertions/changes with respect to reference; the number after the "_" denotes the number of differences to the reference which should correlate with lower expression)
  * E1A_r60: spike-in probes with concentrations that should cover the whole dynamic range of the array

There are a few other expression probes that are used by Agilent's feature extraction/QC pipeline. 

```{r, qc_probes, fig.width=12, fig.height=3}
control_probes = which(intensities$genes$ControlType!=0)
cp_data = intensities$E[control_probes,]
cp_names = intensities$genes[control_probes,]
selected_controls = ifelse(substr(cp_names$ProbeName,1,4)=="ERCC",F,T)

# control probes
for (i in 1:ncol(cp_data)) {
  boxplot(log2(cp_data[selected_controls,i]) ~ factor(cp_names$ProbeName[selected_controls]),las=2, main=paste("Sample",i), outline=F)
}
```

## Number of detected probes
Signal for probes will be considered detected if the expression of a probe is above the 95% quantile of expression in negative control probes. 
In a typical experiment one would expect about 50% of all genes to be detectable.

```{r, expr_count}
neg95 <- apply(normalized$E[normalized$genes$ControlType==-1,],2,function(x) quantile(x,p=0.95, na.rm=T))
cutoff <- matrix(1.1*neg95,nrow(normalized),ncol(normalized),byrow=TRUE)
isexpr <- rowSums(normalized$E > cutoff, na.rm=T) >= 2
# about 50% of genes should be detected
#table(isexpr)
```

```{r, number_expressed,  fig.height=8}
isexpr_sample <- colSums(normalized$E > cutoff, na.rm=T)
sample_order = order(ed[names(isexpr_sample),]$Condition, ed[names(isexpr_sample),]$ShortID)
isexpr_sample_ordered = isexpr_sample[sample_order]
nn = ed[names(isexpr_sample),]$ShortID[sample_order]
par(mar=c(15,4,4,2))
barplot(isexpr_sample_ordered, main="number of detected probes per sample", las=2, names.arg=nn)
par(mar=c(8,4,4,2))
```

```{r}
ofolder = "./Results/data_store/2019-11-15"
save(ed, intensities, normalized, file=file.path(ofolder, "MB194_micro_array_preprocessed_data.Rdata"))
```

# Software versions

```{r}
sessionInfo()
```

